review/get_review_ids.py

```diff
@@ -0,0 +1,235 @@
+        problem_information.append((TEMPLATE_URL.format(course_id=review_course_id,
+                                    type='problem', xblock_id=problem), correct, attempts))
+    return problem_information
+
+
+def get_vertical(current_course):
+    '''
+    Looks through all the problems a learner has previously loaded and
+    finds their parent vertical. Then randomly selects a single vertical
+    to show the learner.
+
+    Parameters:
+        current_course (CourseLocator): The course the learner is currently in
+
+    Returns a url (str) with the vertical id to render for review.
+    '''
+    user = crum.get_current_user()
+
+    enroll_user_in_review_course(user, current_course)
+
+    store = modulestore()
+    course_usage_key = store.make_course_usage_key(current_course)
+    course_blocks = get_course_blocks(user, course_usage_key)
+
+    vertical_data = set()
+
+    for block_key, state in get_records(user, current_course):
+        if is_valid_problem(store, block_key, state):
+            parent = course_blocks.get_parents(block_key)[0]
+            vertical_id = parent.block_id
+            vertical_data.add(vertical_id)
+            delete_state_of_review_problem(user, current_course, block_key.block_id)
+
+    if not vertical_data:
+        return []
+
+    vertical_to_show = random.sample(vertical_data, 1)[0]
+    review_course_id = REVIEW_COURSE_MAPPING[str(current_course)]
+    return (TEMPLATE_URL.format(course_id=review_course_id,
+                                type='vertical', xblock_id=vertical_to_show))
+
+
+def get_records(user, current_course):
+    '''
+    Generator that yields each applicable record from the Courseware Student
+    Module. Each record corresponds to a problem the user has loaded
+    in the original course.
+
+    Parameters:
+        user (django.contrib.auth.models.User): User object for the current user
+        current_course (CourseLocator): The course the learner is currently in
+
+    Returns:
+        record.module_state_key (opaque_keys.edx.locator.BlockUsageLocator):
+            The locator for the problem
+        state (dict): The state of the problem
+    '''
+    problem_filter = {'student_id': user.id, 'course_id': current_course, 'module_type': 'problem'}
+    for record in StudentModule.objects.filter(**problem_filter):
+        state = json.loads(record.state)
+        # The key 'selected' shows up if a problem comes from a
+        # library content module. These cause issues so we skip this.
+        # Issue: Library content contains problems but the CSM brings up
+        # the library content and not the problems within
+        if 'selected' not in state:
+            yield record.module_state_key, state
+
+
+def enroll_user_in_review_course(user, current_course):
+    '''
+    If the user is not enrolled in the review version of the course,
+    they are unable to see any of the problems. This ensures they
+    are enrolled so they can see review problems.
```

Comment:
What are the security requirements for this?  Are you Ok with learners finding out the course-key for the review-course and going directly to that course in the LMS?

Reply:
This is a known concern and one that I'm not sure how to combat at this time. For right now, I'm hoping that by having the review course as a copy of the actual course, they will be limited in content they can see anyway from the review course. I discussed this with a couple members on my team and came to the conclusion that it poses similar security concerns as a user creating multiple accounts and answering a question multiple times that way.